### PR TITLE
[release/6.0-rc2] Parse date types into UTC by default

### PR DESCRIPTION
**Pull request by captainsafia**
## Description

This PR resolves an issue where `DateTime` and `DateTimeOffset` types were deserialized into the local timezone instead of UTC, resulting in confusing behavior for customers building apps with minimal APIs.

## Customer Impact

Without this change, date-type parameters passed as route or parameter queries in minimal apps would be deserialized into the local timezone on the server by default. This can result in confusing behavior where APIs will report different date time values depending on the timezone of the server the API is running in.

To ensure that minimal apps can handle dates predictably we deserialize dates into UTC by default.
## Regression?
- [ ] Yes
- [X] No

## Risk
- [ ] High
- [ ] Medium
- [X] Low

This change has a low surface area, it will only apply to:
- Minimal apps
- Apps that use `DateTime` and `DateTimeOffset` parameters

The change was validated on unit tests and manually.

Behaviorally, this change is not risky since it mimics what is already done in MVC.

## Verification
- [X] Manual (required)
- [X] Automated

## Packaging changes reviewed?
- [ ] Yes
- [ ] No
- [X] N/A

Addresses https://github.com/dotnet/aspnetcore/issues/36725

### CODE REVIEW

File: src/Shared/ParameterBindingMethodCache.cs [hunk 1]
```diff
@@ -88,11 +88,28 @@ public bool HasTryParseMethod(ParameterInfo parameter)
 
                 if (TryGetDateTimeTryParseMethod(type, out methodInfo))
                 {
+                    Expression dateTimeStyles = type == typeof(DateTime) || type == typeof(DateTimeOffset) ?
+                        Expression.Convert(
+                            Expression.Or(
+                                // We conver the `DateTimeStyles` value to the underlying integer time
```
Comment (campersau) [inline src/Shared/ParameterBindingMethodCache.cs:94]:
```suggestion
                                // We convert the `DateTimeStyles` value to the underlying integer time
```

**Review by halter73 — APPROVED** [review]

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 2]
```diff
@@ -2889,6 +2889,49 @@ public async Task RequestDelegateWithBindAndImplicitBodyRejectsNonJsonContent(bo
             }
         }
 
+        public static IEnumerable<object?[]> DateTimeDelegates
+        {
+            get
+            {
+                string dateTimeParsing(DateTime time) => $"Time: {time.ToString("O")}, Kind: {time.Kind}";
```
Comment (pranavkm) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2896]: I know it doesn't matter, but could we use the the culture specifier here regardless:

```suggestion
                string dateTimeParsing(DateTime time) => $"Time: {time.ToString("O", CultureInfo.InvariantCulture)}, Kind: {time.Kind}";
```

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 3]
```diff
@@ -2889,6 +2889,49 @@ public async Task RequestDelegateWithBindAndImplicitBodyRejectsNonJsonContent(bo
...
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "09/20/2021 16:34:09", "Time: 2021-09-20T16:34:09.0000000Z, Kind: Utc" },
+                    new object?[] { (Func<DateTimeOffset, string>)dateTimeOffsetParsing, "09/20/2021 16:34:09 +00:00", "Time: 2021-09-20T16:34:09.0000000+00:00" },
+                    new object?[] { (Func<DateOnly, string>)dateOnlyParsing, "9/20/2021", "Time: 2021-09-20" },
+                    new object?[] { (Func<TimeOnly, string>)timeOnlyParsing, "4:34 PM", "Time: 16:34:00.0000000" },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(DateTimeDelegates))]
```
Comment (pranavkm) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2912]: I'm generally not a fan of using Theory case for testing what amounts to different code paths here. Could we expand these and write bespoke tests? That way if one of these changes in the future you don't have to figure out how to filter out an exact test case using `dotnet test`. You can always a helper function to re-use code.

Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2912]: Yep -- this is fair. I think it is sufficient to split it up into a set of tests for `DateTime`/`DateTimeOffset` and another set of tests for `DateOnly`/`TimeOnly`.

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 4]
```diff
@@ -2889,6 +2889,91 @@ public async Task RequestDelegateWithBindAndImplicitBodyRejectsNonJsonContent(bo
...
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "2021-09-20 4:18:44", "Time: 2021-09-20T04:18:44.0000000, Kind: Unspecified" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "   9/20/2021    4:18:44 PM  ", "Time: 2021-09-20T16:18:44.0000000, Kind: Unspecified" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "2021-09-20T16:28:02.000-07:00", "Time: 2021-09-20T23:28:02.0000000Z, Kind: Utc" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "  2021-09-20T 16:28:02.000-07:00  ", "Time: 2021-09-20T23:28:02.0000000Z, Kind: Utc" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "2021-09-20T23:30:02.000+00:00", "Time: 2021-09-20T23:30:02.0000000Z, Kind: Utc" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "     2021-09-20T23:30: 02.000+00:00 ", "Time: 2021-09-20T23:30:02.0000000Z, Kind: Utc" },
+                    new object?[] { (Func<DateTime, string>)dateTimeParsing, "2021-09-20 16:48:02-07:00", "Time: 2021-09-20T23:48:02.0000000Z, Kind: Utc" },
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2907]: Can we add test cases for DateTimeOffset and verify that `Offset` is always `TimeSpan.Zero`?

File: src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs [hunk 5]
```diff
@@ -2889,6 +2889,91 @@ public async Task RequestDelegateWithBindAndImplicitBodyRejectsNonJsonContent(bo
...
+                    new object?[] { (Func<DateOnly, string>)dateOnlyParsing, "9 /20 /2021", "Time: 2021-09-20" },
+                    new object?[] { (Func<TimeOnly, string>)timeOnlyParsing, "4:34 PM", "Time: 16:34:00.0000000" },
+                    new object?[] { (Func<TimeOnly, string>)timeOnlyParsing, "    4:34 PM   ", "Time: 16:34:00.0000000" },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(DateOnlyDelegates))]
+        public async Task RequestDelegateCanProcessDateOnlyValues(Delegate @delegate, string inputTime, string expectedResponse)
```
Comment (halter73) [inline src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:2955]: Nit: can we split up the DateOnly and TimeOnly tests or rename the test and MemberData to reference both?

**Review by halter73 — APPROVED** [review]

**Comment by Pilchie** [comment]
This was approved over email.

**Comment by captainsafia** [comment]
@dotnet/aspnet-build Can I get help merging?

**Pull request merged**